basyx/aas/adapter/http.py

```diff
@@ -591,21 +605,38 @@ def _get_shells(self, request: Request) -> Iterator[model.AssetAdministrationShe
     def _get_shell(self, url_args: Dict) -> model.AssetAdministrationShell:
         return self._get_obj_ts(url_args["aas_id"], model.AssetAdministrationShell)
 
-    def _get_submodels(self, request: Request) -> Iterator[model.Submodel]:
-        submodels: Iterator[model.Submodel] = self._get_all_obj_of_type(model.Submodel)
+    def _get_submodels(self, request: Request) -> Tuple[Iterator[model.Submodel], int]:
+        limit = request.args.get('limit', type=int, default=10)
+        cursor = request.args.get('cursor', type=int, default=0)
+        submodels: List[model.Submodel] = list(self._get_all_obj_of_type(model.Submodel))
```

Comment:
This first collects all submodels as a list and then slices the list, removing some submodels that were previously collected. We should use `itertools.islice()` here instead, so we don't even collect all submodels.

Reply:
If I want to filter first like you recommend below, don't I need all Submodels anyway?

Reply:
`_get_all_obj_of_type()` returns an iterator, which by itself doesn't do anything. The `filter()` functions wrap this iterator, but by themselves also don't do anything. Finally, you can apply `itertools.islice()` on the final filter iterator, and iterate the iterator returned by `islice()`, which will then evaluate the iterator one by one. Thus, only on iteration a single element is yielded, checked whether it matches the filtering conditions. The `islice()` iterator wrapper will abort iteration once x elements have been yielded.

When instead using `list(_get_all_obj_of_type())`, all elements are first collected in a list, and filtered afterwards, which can take some time depending on the number of elements, and which is also unnecessary, if we only want a limited amount of submodels.